include/up-cpp/transport/builder/UAttributesBuilder.h

```diff
@@ -0,0 +1,193 @@
+
+/// Copyright (c) 2024 General Motors GTO LLC
+/// 
+/// Licensed to the Apache Software Foundation (ASF) under one
+/// or more contributor license agreements.  See the NOTICE file
+/// distributed with this work for additional information
+/// regarding copyright ownership.  The ASF licenses this file
+/// to you under the Apache License, Version 2.0 (the
+/// "License"); you may not use this file except in compliance
+/// with the License.  You may obtain a copy of the License at
+///
+///   http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing,
+/// software distributed under the License is distributed on an
+/// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+/// KIND, either express or implied.  See the License for the
+/// specific language governing permissions and limitations
+/// under the License.
+/// 
+/// SPDX-FileType: SOURCE
+/// SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+/// SPDX-License-Identifier: Apache-2.0
+///
+
+#ifndef _UATTRIBUTESBUILDER_
+#define _UATTRIBUTESBUILDER_
+
+#include <string>
+#include <spdlog/spdlog.h>
+#include <up-cpp/uuid/factory/Uuidv8Factory.h>
+#include <up-core-api/uattributes.pb.h>
+#include <up-core-api/uuid.pb.h>
+
+namespace uprotocol::utransport {
+
+/// @brief UAttributesBuilder to build UAttributes Objects.
+///
+/// The class provides a fluent interface to build UAttributes objects with the
+/// desired attributes.
+class UAttributesBuilder {
+    private:
```

Comment:
> Do we need the `*Builder` classes? As they are implemented right now, they are not versioned to the protobuf objects they build and, while they can validate data during the construction of one of the protobuf objects, once `build()` is called the user can manipulate the protobuf data any way they would like (necessitating validation again at the point where the data is consumed).

You're correct that most uE developers will _not_ care about the fine details.

FWIW, I think @sophokles73 struck a nice balance when he implemented [`UMessageBuilder`](https://github.com/eclipse-uprotocol/up-rust/blob/main/src/umessage/umessagebuilder.rs) over in `up-rust`. We can choose to build a `Publish`, `Notification`, `Request`, `Response` message, then call methods to layer on additional attributes if necessary or we want non-default. Then we finally call one of the `build*` functions to 

Perhaps making the API cut more at that level here makes sense too?

Reply:
The protobuf project does not recommend ever inheriting directly from protobuf message declarations.
Regarding builder architecture, I have proposed that we have wrapper objects encapsulating the protobuf messages with pImpl (incomplete type declaration of a shared_ptr). I have test code demonstrating this working as a builder pattern, including for nested objects. This pattern also supports fetching from the protobuf messages through the same pattern. The things that are missing is support for 'oneof' (which we make very limited use of) and 'repeated', which I believe we don't use. I have also proposed using python script to autogen all of the wrapping from the protobuf definition. This autogen is not currently aligned with the code demo.

Having dug a bit deeper on how to best support 'oneof', I'm learning of the reflection interface built into protobuf, and becoming of the opinion that autogen is not necessary, or necessary in a much more limited context. The reflection interface was brought to my attention weeks back, probably by Greg, but I did not grasp the value then. The ideal condition is to be able to build a fully dynamic api for creating the serialized attributes from a dll loaded implementation, and skip the builder pattern.

A part of this protobuf issue is the validation. I believe the existing validation code must rest on the protobuf validator, and least in part. The ideal circumstance would be to have all of the validation occur in a single call to the protobuf validator for the top level protobuf object. For this to work, I would suspect the protobuf specs would have to be significantly improved to be precise enough. It would be desirable to have dynamic versioning support to be able to allow protobuf extension, if this path were taken.

Reply:
The builders that remain have been streamlined with interfaces focused on specific use scenarios. Validators have been extended such that they can be used to validate protobuf objects received from callers at the library interfaces, and exceptions will be thrown if an interface is called with invalid or inappropriate data.

Additionally, L2 clients have been added that further support (and limit) specific use cases like pub/sub, notifications, and RPC.